### PR TITLE
Create getProposedFederatorPublicKeyOfType in Federation Support

### PR DESCRIPTION
**Pull request by apancorb**
## Description
Create getProposedFederatorPublicKeyOfType in Federation Support.

## Motivation and Context
https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP419.md

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java [hunk 1]
```diff
@@ -374,6 +374,13 @@ public byte[] getPendingFederatorPublicKeyOfType(int index, FederationMember.Key
     public Optional<Federation> getProposedFederation() {
         return provider.getProposedFederation(constants, activations);
     }
+    
+    @Override
+    public Optional<byte[]> getProposedFederatorPublicKeyOfType(int index, FederationMember.KeyType keyType) {
+        return getProposedFederation()
+            .map(Federation::getMembers)
+            .map(members -> getFederationMemberPublicKeyOfType(members, index, keyType, "Proposed Federation"));
```
Comment (julia-zack) [inline rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java:382]:
```suggestion
            .map(members -> getFederationMemberPublicKeyOfType(members, index, keyType, "Proposed Federator"));
```

:)

File: rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java [hunk 2]
```diff
@@ -802,17 +809,20 @@ public long getActiveFederationCreationBlockHeight() {
     }
 
     /**
-     * Returns the compressed public key of given type of the member list at the given index
-     * Throws a custom index out of bounds exception when appropiate
-     * @param members the list of federation members
-     * @param index the federator's index (zero-based)
-     * @param keyType the key type
+     * Returns the compressed public key of given type of the member list at the
+     * given index.Throws a custom index out of bounds exception when appropiate.
```
Comment (julia-zack) [inline rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java:813]:
```suggestion
     * given index. Throws a custom index out of bounds exception when appropiate.
```

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 3]
```diff
@@ -2209,15 +2208,41 @@ void getProposedFederation_whenStorageProviderReturnsProposedFederation_shouldRe
...
+        assertFalse(actualPublicKey.isPresent());
+    }
+
+    @Test
+    void getProposedFederatorPublicKeyOfType_whenStorageProviderReturnsProposedFederation_shouldReturnPublicKey() {
+        // Arrange
+        Federation proposedFederation = P2shErpFederationBuilder.builder().build();
+        storageProvider.setProposedFederation(proposedFederation);
+
+        FederationMember.KeyType keyType = FederationMember.KeyType.BTC;
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2226]: lets make tests for RSK and MST keys too

Comment (apancorb) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2226]: Sounds good, will update

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 4]
```diff
@@ -2209,15 +2208,41 @@ void getProposedFederation_whenStorageProviderReturnsProposedFederation_shouldRe
...
+        byte[] expectedPublicKey = proposedFederation.getMembers().get(0).getPublicKey(keyType).getPubKey(true);
```
[the remark this answers is not in the document]
Comment (apancorb) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2227]: Isn't this the same?

File: rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java (hunk 2 above)
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java]: Updated

File: rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java (hunk 1 above)
Comment (apancorb) [reply rskj-core/src/main/java/co/rsk/peg/federation/FederationSupportImpl.java]: Updated

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 3 above)
Comment (apancorb) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java]: Updated

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (julia-zack) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2272]: just to be explicit about the pub key type, but no worries

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 5]
```diff
@@ -2232,15 +2231,86 @@ void getProposedFederationAddress_whenStorageProviderReturnsProposedFederation_s
...
+    }
+
+    @Test
+    void getProposedFederatorPublicKeyOfType_whenStorageProviderReturnsProposedFederationAndKeyTypeIsBTC_shouldReturnPublicKey() {
+        // Arrange
+        List<BtcECKey> federationKeys = BitcoinTestUtils.getBtcEcKeysFromSeeds(
+            new String[] { "fa01", "fa02", "fa03", "fa04", "fa05", "fa06", "fa07", "fa08", "fa09" }, true);
+        Federation proposedFederation = P2shErpFederationBuilder.builder().withMembersBtcPublicKeys(federationKeys).build();
+        FederationMember.KeyType keyType = FederationMember.KeyType.BTC;
+        byte[] expectedPublicKey = proposedFederation.getMembers().get(0).getPublicKey(keyType).getPubKey(true);
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2250]:
```
BtcECKey publicKey1 = BtcECKey.fromPublicOnly(Hex.decode("0261631efa5f32602a3e2b1dc89d648c3212f17fd87e26614cb3a1d9516346098a");
List<BtcECKey> federationPublicKeys = Arrays.asList(
            publicKey1,
            BtcECKey.fromPublicOnly(Hex.decode(""),
            etc
        );
```

Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2250]:
```suggestion
        BtcECKey publicKey1 = BtcECKey.fromPublicOnly(Hex.decode("0261631efa5f32602a3e2b1dc89d648c3212f17fd87e26614cb3a1d9516346098a");
        List<BtcECKey> federationKeys = Arrays.asList(
            publicKey1,
            BtcECKey.fromPublicOnly(Hex.decode(""),
            etc
        );
        Federation proposedFederation = P2shErpFederationBuilder.builder().withMembersBtcPublicKeys(federationKeys).build();

        FederationMember.KeyType keyType = FederationMember.KeyType.BTC;
        byte[] expectedPublicKey = publicKey1.getPubKey(true);
```

wdyt?

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 6]
```diff
@@ -2309,6 +2309,77 @@ void getProposedFederationCreationBlockNumber_whenStorageProviderReturnsProposed
         assertEquals(creationBlockNumber, actualCreationBlockNumber.get());
     }
 
+    @Test
+    void getProposedFederatorPublicKeyOfType_whenFederationIsEmpty_shouldReturnEmpty() {
```
Comment (marcos-iov) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2313]:
```suggestion
    @ParameterizedTest
    @ValueSource(KeyType.BTC, KeyType.RSK, KeyType.MST)
    void getProposedFederatorPublicKeyOfType_whenFederationIsEmpty_shouldReturnEmpty() {
```
What do you think? To cover some extra scenarios. Same can be applied to the remaining tests

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 5 above)
Comment (apancorb) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2328]: For me it is as explicit as it should be. Open to more thoughts/ideas.

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 6 above)
Comment (apancorb) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java]: Sounds good, I have updated it. Let me know if it looks fine.

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java [hunk 7]
```diff
@@ -2309,6 +2310,78 @@ void getProposedFederationCreationBlockNumber_whenStorageProviderReturnsProposed
...
+    }
+
+    @Test
+    void getProposedFederatorPublicKeyOfType_whenStorageProviderReturnsProposedFederationAndKeyTypeIsBTC_shouldReturnPublicKey() {
+        // Arrange
+        List<BtcECKey> federationKeys = BitcoinTestUtils.getBtcEcKeysFromSeeds(
+            new String[] { "fa01", "fa02", "fa03", "fa04", "fa05", "fa06", "fa07", "fa08", "fa09" }, true);
+        Federation proposedFederation = P2shErpFederationBuilder.builder().withMembersBtcPublicKeys(federationKeys).build();
+        FederationMember.KeyType keyType = FederationMember.KeyType.BTC;
+        byte[] expectedPublicKey = proposedFederation.getMembers().get(0).getPublicKey(keyType).getPubKey(true);
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2330]:
```suggestion
byte[] expectedPublicKey = federationKeys.get(0).getPubKey();
```
according to the discussion we had, right?

File: rskj-core/src/test/java/co/rsk/peg/federation/P2shErpFederationBuilder.java [hunk 8]
```diff
@@ -96,18 +94,21 @@ public ErpFederation build() {
 
     private List<FederationMember> getFederationMembers() {
         if (membersRskPublicKeys.isEmpty()) {
-            membersRskPublicKeys = membersBtcPublicKeys.stream()
-                .map(btcKey -> ECKey.fromPublicOnly(btcKey.getPubKey()))
-                .collect(Collectors.toList());
+            this.membersRskPublicKeys = membersBtcPublicKeys.stream()
+                .map(BtcECKey::getPubKey)
+                .map(ECKey::fromPublicOnly)
+                .toList();
```
Comment (julia-zack) [inline rskj-core/src/test/java/co/rsk/peg/federation/P2shErpFederationBuilder.java:100]: nice

File: rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java (hunk 7 above)
Comment (marcos-iov) [reply rskj-core/src/test/java/co/rsk/peg/federation/FederationSupportImplTest.java:2330]: The order might be different, first key in `federationKeys` might not match with the first member. Keys are sorted by hex value

**Review by marcos-iov — APPROVED** [review]

**Pull request merged**